packages/@uppy/companion-client/types/index.d.ts

```diff
@@ -10,20 +10,31 @@ export interface TokenStorage {
   removeItem: (key: string) => Promise<void>
 }
 
+type CompanionHeaders = Record<string, unknown>
 export interface RequestClientOptions {
   companionUrl: string
-  companionHeaders?: Record<string, unknown>
+  companionHeaders?: CompanionHeaders
   companionCookiesRule?: RequestCredentials
 }
 
+// TODO: remove boolean support for options that was added for backward compatibility.
+type RequestOptions = {
+  skipPostResponse?: boolean,
+  signal?: AbortSignal,
+} | boolean
+
 export class RequestClient {
   constructor (uppy: Uppy, opts: RequestClientOptions)
 
-  get (path: string): Promise<any>
+  readonly hostname: string
+
+  setCompanionHeaders(headers: CompanionHeaders): void
+
+  get<T = unknown> (path: string, options?: RequestOptions): Promise<T>
 
-  post (path: string, data: Record<string, unknown>): Promise<any>
+  post<T = unknown> (path: string, data: Record<string, unknown>, options?: RequestOptions): Promise<T>
```

Comment:
We don't want a generic, we want to do [method overloading](https://www.typescriptlang.org/docs/handbook/2/functions.html#function-overloads)

Reply:
I think we are not talking about the same thing, we don't know what the function will return because it will depend what the network request is sending back, I don't think it has anything to do with overloading.

Reply:
I thought you wanted to support both boolean option and object option as in the code, for which you need overloading. I think we should do that regardless.

Regarding the generic for the return type, now that I understand it makes sense indeed